builtin/credential/aws/path_role.go

```diff
@@ -348,14 +399,127 @@ func (b *backend) pathRoleCreateUpdate(
 		roleEntry.BoundIamInstanceProfileARN = boundIamInstanceProfileARNRaw.(string)
 	}
 
-	// Ensure that at least one bound is set on the role
-	switch {
-	case roleEntry.BoundAccountID != "":
-	case roleEntry.BoundAmiID != "":
-	case roleEntry.BoundIamInstanceProfileARN != "":
-	case roleEntry.BoundIamRoleARN != "":
-	default:
+	if boundIamPrincipalARNRaw, ok := data.GetOk("bound_iam_principal_arn"); ok {
+		roleEntry.BoundIamPrincipalARN = boundIamPrincipalARNRaw.(string)
+	}
+
+	if inferRoleTypeRaw, ok := data.GetOk("role_inferred_type"); ok {
+		roleEntry.RoleInferredType = inferRoleTypeRaw.(string)
+	}
+
+	if inferredAWSRegionRaw, ok := data.GetOk("inferred_aws_region"); ok {
+		roleEntry.InferredAWSRegion = inferredAWSRegionRaw.(string)
+	}
+
+	var allowEc2Auth, allowIamAuth bool
+
+	parseAllowedAuthTypes := func(input string) string {
+		allowedAuthTypes := []string{}
+		for _, t := range strings.Split(input, ",") {
+			switch t {
+			case "ec2":
+				allowedAuthTypes = append(allowedAuthTypes, t)
+			case "iam":
+				allowedAuthTypes = append(allowedAuthTypes, t)
+			case "":
+			default:
+				return fmt.Sprintf("unrecognized auth type: '%s'", t)
+			}
+		}
+		roleEntry.AllowedAuthTypes = allowedAuthTypes
+		return ""
+	}
+
+	if allowedAuthTypesRaw, ok := data.GetOk("allowed_auth_types"); ok {
+		err := parseAllowedAuthTypes(allowedAuthTypesRaw.(string))
+		if err != "" {
+			return logical.ErrorResponse(err), nil
+		}
+	} else if req.Operation == logical.CreateOperation {
+		err := parseAllowedAuthTypes(data.Get("allowed_auth_types").(string))
+		if err != "" {
+			return logical.ErrorResponse(err), nil
+		}
+	}
+
+	// Reparse the existing roleEntry.AllowedAuthTypes
+	// We need to do this to support the use case where an existing role is being updated, and
+	// allowed_auth_types isn't being updated as part of the role. We need to make sure that any
+	// new bindings requested are still valid bindings. For example, let's say a role is created
+	// with an auth_type of iam and no inferencing is configured; then, in a subsequent role update,
+	// bound_ami_id is specified. This is how that edge case is caught.
+	for _, t := range roleEntry.AllowedAuthTypes {
```

Comment:
I see a possibility of only doing this once. Basically, retain this block as-is and change the above code as follows:

```
 if allowedAuthTypesRaw, ok := data.GetOk("allowed_auth_types"); ok {
		roleEntry.AllowedAuthTypes = strutil.ParseDedupAndSortStrings(allowedAuthTypesRaw.(string), ",")
	} else if req.Operation == logical.CreateOperation {
		roleEntry.AllowedAuthTypes = strutil.ParseDedupAndSortStrings(data.Get("allowed_auth_types").(string), ",")
	}
```

Let me know if it doesn't address all the use cases you intended to address.